### PR TITLE
[CR]Pain nerf

### PR DESCRIPTION
**Pull request by Coolthulhu**
There is a big problem with pain: it kills long before damage does.

This could be solved in multiple ways, but neither of them is perfect.
I went with sub-linear scaling.

Old formula was `pain/15`, new one is `(pain^0.8)/10`.
Pain breakpoints look like this:

| Str/dex loss | 1 | 2 | 3 | 4 | 5 | 6 |
| --- | --- | --- | --- | --- | --- | --- |
| Old pain | 15 | 30 | 45 | 60 | 75 | 90 |
| New pain | 18 | 43 | 71 | 101 | 133 | 167 |

Speed loss is also changed from `pain / 2` to `pain^0.7`. Changed cap from 60 to 50.

| Pain | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 100 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Old penalty | 5 | 10 | 15 | 20 | 25 | 30 | 35 | 40 | 45 | 50 |
| New penalty | 5 | 8 | 10 | 13 | 15 | 17 | 19 | 21 | 23 | 25 |

Other changes:
- Removed pain being rounded down, causing small hits to cause no pain. Instead remainders are rolled now.
- Removed really weird extra pain formula for cut and stab damage. It was `x+sqrt(x)`. The difference is minimal.
- Lowered pain from bashing attacks to compensate for pain flooring. Was `dmg/4`, is `dmg/5`.
- Removed special handling of cold damage. Nothing deals cold damage, meaning that it would need to be reimplemented anyway.
- Calculated pain penalties in one function that returns a struct. Results in a bit of redundancy, but better than having same function repeated in 2 places.

### CODE REVIEW

**Comment by pisskop** [comment]
speed loss killed long before stat loss,  the old stat loss system was fine.

**Comment by Coolthulhu** [comment]
Stat loss kills dodge and strength (causing weight penalties), meaning it is quite deadly to weak characters.

**Comment by mugling** [comment]
> There is a big problem with pain: it kills long before damage does.
> 
> This could be solved in multiple ways, but neither of them is perfect.
> I went with sub-linear scaling.

Could we tackle this problem at source - make application of pain sub-linear and/or stop it being strictly additive (in the case of the latter stopping small amounts 'stacking' so much)

**Comment by pisskop** [comment]
There is an arguement to be made that weak characters shouldnt be carrying so much that they cant escape a combat situation, or that they can just drop the excess.

The reason why you take more strength is to provide a buffer as much as it is to carry heavier armor around.  If i can take an example from another game; morrowind.  You took 100 strength to carry daedric.  At 5 pounds per strength point, you naturally maxed out at 500 pounds load.  A full suit of daedric weighed 450 pounds, leaving 50 for weapons and potions and ingredients and question items and other knicknacs.  The light armor left something like 200 pounds to carry other nonarmor items.


The same principle applies here.  You take higher strength to carry more armor and heavier weapons.

**Comment by mugling** [comment]
> Could we tackle this problem at source - make application of pain sub-linear and/or stop it being strictly additive (in the case of the latter stopping small amounts 'stacking' so much)
> 
> This has the problem of order of application: getting slapped a few times and then getting smashed once would have different resulting pain values.

Didn't think of that. Possibly some system of adding (or replacing the existing value) only if above a threshold?

**Comment by Coolthulhu** [comment]
> Possibly some system of adding (or replacing the existing value) only if above a threshold?

That would cause debuff spikes after getting smashed, which is already an emergency.

**Comment by mightyagrippa** [comment]
One possibility, if keeping track of player sources of pain separately is possible. Keep track of player's highest source of pain and reduce any sources of pain less than some % of that. For example, if threshold is set at 20%, getting slammed through a wall for 40 pain will dull the pain of any wounds causing 8 or less pain. 

There would have to be a cap on the lessening effect. Without it there are some weirdnesses, like if zombie children nibble you to 100 pain then getting smashed through a wall could actually reduce your pain.

If tracking pains separately is possible you could also exponentially reduce pain to prevent horrible stacking. Something like the nth biggest source of pain is multiplied by .9^(n-1), cutting the 7th largest pain source in half, 14th in quarter, etc. Or truncate the list after a certain point to get a similar stacking prevention.

**Comment by mugling** [comment]
> One possibility, if keeping track of player sources of pain separately is possible

I like this a lot but it's quite a bit of work and it is therefore best to go with an implementation that currently exists and is a nett improvment. You could open an issue to track this suggestion further though.

File: src/creature.cpp [hunk 1]
```diff
@@ -724,46 +724,40 @@ void Creature::deal_damage_handle_type(const damage_unit &du, body_part bp, int
...
     }
+
+    float div = 4.0f;
+
+    switch( du.type ) {
+        case DT_BASH:
+            // Bashing damage is less painful
+            div = 5.0f;
+            break;
+        case DT_CUT:
```
Comment (mugling) [inline src/creature.cpp]: these should be handled by `default`

**Comment by mugling** [comment]
Separate issue - when did armor get massively improved?

**Comment by Rivet-the-Zombie** [comment]
>when did armor get massively improved?

I wasn't aware it had been improved. What's different?

**Comment by mugling** [comment]
I noticed it a while ago but this does magnify the effect a little. Start as a janitor. Stand in a horde of zombies. Notice how little damage they actually do - previously the pain was what killed you (fixed by this PR). You can hold down pause for a looong time and still live

**Comment by Coolthulhu** [comment]
I don't recall any recent, intentional change to armor.
Could be a bug, dodge, rng or just the fact that most zombies have horribly low damage.

**Comment by mugling** [comment]
Any of the above excluding rng - too consistently observed

**Comment by pisskop** [comment]
Zombies, doing 2d3 Bashing, can be defeated by as little as 3 undershirts.

**Comment by mightyagrippa** [comment]
Don't forget the bite attack on a 5 turn cooldown, does something like 1d4+5 damage.

**Comment by pisskop** [comment]
this also jacked up how much pain attacks deal?

**Comment by Coolthulhu** [comment]
Removed rounding down, meaning grazes matter now.

**Pull request merged**